LiS/man/man2/putpmsg.2s.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id: putpmsg.2s.man,v 1.1.4.4 2008-04-29 08:33:17 brian Exp $
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date: 2008-04-29 08:33:17 $ by $Author: brian $
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log: putpmsg.2s.man,v $
.\" Revision 1.1.4.4  2008-04-29 08:33:17  brian
.\" - update headers for Affero release
.\"
.\" Revision 1.1.4.3  2006/04/05 09:42:18  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH PUTPMSG 2 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ System Calls"
.SH NAME
.B putpmsg
\- put a band message to a STREAMS character device
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <stropts.h>
.HP 8
.BI "int " retval " = putpmsg(int " fd ", const struct strbuf *" ctlptr ", const struct strbuf *" dataptr ", int " band ", int " flags );
.\"
.\"
.SH ARGUMENTS
.PP
.TP
.I fd
the file descriptor to which to put the message.
.TP
.I ctlptr
a pointer to a
.B strbuf
structure describing the control part of the message.
.TP
.I dataptr
a pointer to a
.B strbuf
structure describing the data part of the message.
.TP
.I band
the band for the message.
.TP
.I flags
flags for the message.
.\"
.\"
.SH DESCRIPTION
.PP
.B putpmsg()
generates a STREAMS message from the buffer supplied in the specified
.IR ctlptr " and " datapr
for the queue band
.I band
and with options specified by
.I flags
and delivers the message to the stream head associated with the STREAMS
character special file
.IR fd .
.PP
The resulting STREAMS message can contain a control part as specified by
.IR ctlptr ,
a data part as specified by
.IR datptr ,
or both.
The control part, when present, will generate
.BR M_PROTO " or " M_PCPROTO
message blocks.
The data part, when present, will generate
.B M_DATTA
message blocks.
When both control and data parts are present, an
.BR M_PROTO " or " M_PCPROTO
message block followed by one or more
.B M_DATA
message blocks will be generated.
.PP
.I fd
is an open file descriptor for the STREAMS character special file to which the
STREAMS message is to be written.
.PP
.IR ctlptr " and " datptr
point to a
.B strbuf
structure, which contains the following members:
.sp
.nf
\fC\s-2\
struct strbuf {
    int maxlen;         /* Maximum buffer length.  */
    int len;            /* Length of data.  */
    char *buf;          /* Pointer to buffer.  */
};
\s+2\fR
.fi
.TP 12
.I maxlen
Not used by
.BR putpmsg() .
.TP 12
.I len
The length of the information pointed to by
.IR buf .
.TP 12
.I buf
A pointer to the buffer that contains
.I len
bytes of information.
.PP
.I ctlptr
points to a
.B strbuf
structure describing the control part to be included in the message, or
.B NULL
if there is no control part.
The
.I buf
member
of the
.B strbuf
structure pointed to by
.I ctlptr
indiates the start of the information to sent in the control part, and
the
.I len
member indicates the length of the information for the control part.
.PP
.I datptr
points to a
.B strbuf
structure decribing the data part to be included in the message, or
.B NULL
is there is no data part.
The
.I buf
member
of the
.B strbuf
structure pointed to by
.I ctlptr
indiates the start of the information to sent in the data part, and
the
.I len
member indicates the length of the information for the data part.
.I len
can be zero (0) to send zero-length data.
.PP
.I band
specifies the queue band to which to write data when
.I flags
is set to
.BR MSG_BAND .
When flags is set to other than
.BR MSG_BAND ,
.I band
must be set to zero (0).
.PP
.I flags
specifies whether to send a high priority message or a band message.
.I flags
can be one of the following values:
.TP 12
.B MSG_HIPRI
Specifies that a high priority message is to be sent if the request contains a
control part
.IR "" ( ctlptr " is not"
.BR "" NULL ).
This results in an
.B M_PCPROTO
message being placed on the stream head.
.TP 12
.B MSG_BAND
Specifies that a normal priority queue band message is to be sent for the
specified
.IR band .
This results in an
.B M_PROTO
message being placed on the stream head when there is a control part.
When there is only a data part, an
.B M_DATA
message for the specified
.I band
is placed on the stream head.
.\"
.\"
.SH "RETURN VALUE"
.PP
Upon success,
.B putpmsg()
returns zero (0).
.PP
Upon failure,
.B putpmsg()
returns 1 and sets
.I errno
to an appropritate error number.
.\"
.\"
.SH ERRORS
.PP
When
.B putpmsg()
fails, it returns -1 and sets
.I errno
to one of the following:
.TP 12
.\".RB [ EACCES ]
.\".I fd
.\"is a dynamic device, and write permission on the device is denied.
.TP 12
.RB [ EAGAIN ]
The stream head is set for non-blocking operation and
.B putpmsg()
would block: that is, the message is being sent with
.I flags
set to
.B MSG_BAND
and the corresponding
.I band
is flow controlled.
.TP 12
.RB [ EBADF ]
.I fd
is not a valid file descriptor or is not open for reading.
.TP 12
.RB [ EFAULT ]
.IR ctlptr ", " datptr ", " ctlptr->buf " or " datptr->buf
point outside the caller's address space.
.TP 12
.RB [ EINTR ]
A signal arrived before the operation could complete.
.TP 12
.RB [ EINVAL ]
The values of
.IR band " or " flags ,
or the combination of
.IR ctlptr ", " datptr ", and the values of " band " and " flags
were invalid,
such as when
.I flags
is specified as
.B MSG_HIPRI
and
.I band
is non-zero,
or,
.I fd
refers to a stream that is linked under a STREAMS multiplexing driver.
.TP 12
.RB [ EIO ]
I/O error, or,
.I fd
refers to a stream that is open but is in the process of closing, or refers to
a pipe and the other end of the pipe is closed.
.TP 12
.RB [ EISDIR ]
.I fd
refers to a directory.
.TP 12
.RB [ ENODEV ]
.I fd
refers to a device that does not support the
.B putpmsg()
system call.
.TP 12
.RB [ ENOSR ]
STREAMS resources were insufficient to complete the operation.
.TP 12
.RB [ ENOSTR ]
.I fd
refers to a device that has no stream associated with it.
.TP 12
.RB [ ENXIO ]
A
.B M_HANGUP
message arrived at the stream head for the specified
.IR fd .
.TP 12
.RB [ ERANGE ]
The
.I len
of the data part of the message is outside the range
.I q_minpsz
to
.I q_maxpsz
inclusive for the topmost STREAMS module or driver under the stream head, or,
the
.I len
of the control or data part is larger than the maximum sized control or data
message for the STREAMS subsystem.
.PP
Other errors may be returned by
.BR putpmsg() .
If the protocol module sends an
.B M_ERROR
message to the stream head, the error returned on all subsequent write
operations, including
.BR putpsmg() ,
is specified in the
.B M_ERROR
message by the protocol module.  Which errors are returned under what
conditions form part of the necessary documentation of the protocol module.
.\"
.\"
.SH CAVEATS
.PP
.B LiS
.[
LiS
.]
could handle concurrent writes better.  Linux socket code blocks concurrent
writers independently while partial writers are installed agains a socket.
When partial writes are performed for messages using
.BR putpmsg() ,
.B LiS
.[
LiS
.]
should install the partial writer and block all writers from completing the
partial write on a given band until the partial writer completes the write or
the partial writer closes the stream.  This would at least keep multiple
threads of execution from interfering with partial writes.
.\"
.\"
.SH BUGS
.PP
.B LiS
.[
LiS
.]
incorrectly returns
.RB [ EINVAL ]
instead of
.RB [ EIO ]
when the internal head queue pointer points to a structure that is not a queue
structure, or is mangled.
.PP
.B LiS
.[
LiS
.]
does not return
.RB [ EIO ]
when
.B putpmsg()
is called on a closing stream.
.PP
.B LiS
.[
LiS
.]
incorrectly returns
.RB [ EAGAIN ]
instead of
.RB [ EINTR ]
when an
.B M_SIG
message causes a signal that interrupts the call to
.B putpmsg()
before any data is written.
.PP
.B LiS
.[
LiS
.]
incorrectly returns
.RB [ ENODEV ]
instead of
.RB [ ENOSTR ]
when there is no stream associated with
.IR fd .
.PP
.B LiS
.[
LiS
.]
incorrectly returns
.RB [ EINVAL ]
instead of
.RB [ ENODEV ]
when
.I fd
refers to a device that does not support the
.B putpmsg()
system call.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR intro (2),
.BR poll (2),
.BR putmsg (2),
.BR write (2),
.BR streamio (2).
.\"
.\"
.SH "CONFORMING TO"
.PP
.BR SVID ", " SUSv2 ", " SUSv3 ", " POSIX .
.\"
.\"
.SH VERSIONS
.PP
This manpage was written for
.B LiS\c
.
.[
LiS
.]
.PP
.B putpmsg()
first appeared in SVR 3.
.[
svr3
.]
.\"
.\"
.XX
.[
magic
.]
.[
svr42
.]
.[
svr4
.]
.[
svr3
.]
.[
svid
.]
.[
xns
.]
.[
xbd5
.]
.[
$LIST$
.]
.TI


NEWS
OpenSS7 Master Package -- history of user visible changes.  2008-05-03
$Id: NEWS,v 0.9.2.53 2008-05-03 21:22:25 brian Exp $
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2000  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
OpenSS7 Master Package package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Documented `lib32gcc1' problem on Ubuntu.

   - Noted problem running under XEN kernels.  XEN kernels are not yet
     supported.

   - License upgrade to AGPL v3.

   - Cross-compile fixes (strap out AC_FUNC_REALLOC macro when
     cross-compiling).

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.

   - Improvements to common build process for embedded and
     cross-compile targets.

   - Support for flex 2.5.33  in maintainer mode.

   - Added the following sub-packages to the master build (and relase):

         - `strx25-0.9.2.1'


_This is a public stable production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
RPM- and DPKG-based distributions, and can be used on production
kernels without patching or recompiling the kernel.

This package is publicly released under the `GNU Affero General Public
License Version 3'.  The release is available as an `autoconf' tarball,
SRPM, DSC, and set of binary RPMs and DEBs.  See the downloads page
(http://www.openss7.org/download.html) for the `autoconf' tarballs,
SRPMs and DSCs.  For tarballs, SRPMs, DSCs and binary RPMs and DEBs,
see the openss7 package page (http://www.openss7.org/openss7_pkg.html).

See `http://www.openss7.org/codefiles/openss7-0.9.2.G/ChangeLog' and
`http://www.openss7.org/codefiles/openss7-0.9.2.G/NEWS' in the release
for more information.  Also, see the `openss7.pdf' manual in the
release (also in html `http://www.openss7.org/openss7_manual.html').

 Major changes for release openss7-0.9.2.F
------------------------------------------

This is the fifth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Support build on openSUSE 10.2.

   - Support build on Fedora 7 and 2.6.21 kernel.

   - Support build on CentOS 5.0 (RHEL5).

   - Support build on Ubuntu 7.04.

   - Updated to gettext 0.16.1.

   - Changes to support build on 2.6.20-1.2307.fc5 and
     2.6.20-1.2933.fc6 kernel.

   - Supports build on Fedora Core 6.

   - Support for recent distributions and tool chains.


 Major changes for release openss7-0.9.2.E
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

It is unlikely that the `OpenSS7 Master Package' will be released as
frequently as before.  Sub-packages will be released more often.  To
rebuild the master package with a new sub-package release, simply
replace the directory to which the sub-package belongs with the
unpacked sub-package release and then rebuild the master package.  This
release provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Addition of the `osr61' sub-package that contains Dialogic(R) Open
     System Release 6.1 version 239 GPL drivers.

   - A few minor corrections to the common build process.

   - Support for autoconf 2.61, automake 1.10 and gettext 0.16.

   - Support for Ubuntu 6.10 distribution and bug fixes for i386
     kernels.

   - The package now looks for subpackages with a version number as
     unpacked by separate tarball.


 Major changes for release openss7-0.9.2.D
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  The
sub-packages have been reorganized for this release.  See `README' in
the release for a sub-package listing.  Aside from sub-package
reorganization, the major difference from previous release is that this
release no longer contains `LiS'.  Too many of the sub-packages will
not even build against `LiS' because of its Stream head deficiencies.


LiS/libc/putpmsg.c
/*****************************************************************************

 @(#) $RCSfile: putpmsg.c,v $ $Name:  $($Revision: 1.1.1.3.4.11 $) $Date: 2008-04-29 08:33:16 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-04-29 08:33:16 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: putpmsg.c,v $
 Revision 1.1.1.3.4.11  2008-04-29 08:33:16  brian
 - update headers for Affero release

 Revision 1.1.1.3.4.10  2007/08/14 10:47:13  brian
 - GPLv3 header update

 *****************************************************************************/

#ident "@(#) $RCSfile: putpmsg.c,v $ $Name:  $($Revision: 1.1.1.3.4.11 $) $Date: 2008-04-29 08:33:16 $"

static char const ident[] =
    "$RCSfile: putpmsg.c,v $ $Name:  $($Revision: 1.1.1.3.4.11 $) $Date: 2008-04-29 08:33:16 $";

/* This file can be processed with doxygen(1). */

#include "streams.h"

/** @addtogroup strcalls STREAMS System Calls
  * @{ */

/** @file
  * STREAMS System Call putpmsg() implementation file.  */

#ifdef BLD32OVER64
/**
 * @struct strbuf6
 * @date 2006-02-20
 * @author bidulock@openss7.org
 *
 * This is really, really, the wrong way to do this.  I cannot stress how wrong
 * this is.  This does not provide 32/64 bit compatibility at all.  It simply
 * attempts to turn a 32 bit call into a 64 bit call an passes it to the kernel.
 * A big problem is the pointer conversion (32->64), which is architecture
 * specific.  (Some need (void *)(long)(int32_t), some need (void
 * *)(ulong)(uint32_t), others might need to supply an offset.) For example of
 * how wrong this is: applications built on s390 using static libraries will not
 * run on s390x.  The proper way to do this is as it is done under Linux
 * Fast-STREAMS: leave the library alone and handle the conversion in the
 * kernel.  When performed by syscall, the syscall32 interface must perform the
 * conversion.  When performed by read/write call, the "magic length" must
 * indicate whether conversion is necessary.  When performed by ioctl, the 32/64
 * bit ioctl conversion functions must do the job.  I really have no desire to
 * "fix" LiS in this regard.  It is deprecated: don't use it.  Use Linux
 * Fast-STREAMS instead.
 */
#endif

#ifdef BLD32OVER64
typedef struct strbuf6 {
	int maxlen;
	int len;
	unsigned long long buf;
} strbuf6_t;

typedef struct putpmsg_args6 {
	int fd;
	unsigned long long ctl;
	unsigned long long dat;
	int band;
	int flags;
} putpmsg_args6_t;
#endif

/** @brief put a message to a stream band.
  * @param fd a file descriptor representing the stream.
  * @param ctlptr a pointer to a strbuf structure describing the control part of the message.
  * @param datptr a pointer to a strbuf structure describing the data part of the message.
  * @param band the band to which to put the message.
  * @param flags the priority of the message.
  */
int
__old_lis_putpmsg(int fd, struct strbuf *ctlptr, struct strbuf *datptr, int band, int flags)
{
#ifdef BLD32OVER64
	putpmsg_args6_t args;
	strbuf6_t ctl6;
	strbuf6_t *ptrc6 = 0;
	strbuf6_t dat6;
	strbuf6_t *ptrd6 = 0;

	if (ctlptr) {
		ctl6.maxlen = ctlptr->maxlen;
		ctl6.len = ctlptr->len;
		ctl6.buf = (unsigned long long) (unsigned long) ctlptr->buf;
		ptrc6 = &ctl6;
	}

	if (dataptr) {
		dat6.maxlen = dataptr->maxlen;
		dat6.len = dataptr->len;
		dat6.buf = (unsigned long long) (unsigned long) dataptr->buf;
		ptrd6 = &dat6;
	}

	memset((void *) &args, 0, sizeof(putpmsg_args6_t));
	args.fd = fd;
	args.ctl = (unsigned long long) (unsigned long) ptrc6;
	args.dat = (unsigned long long) (unsigned long) ptrd6;
	args.band = band;
	args.flags = flags;

	return (write(fd, &args, LIS_GETMSG_PUTMSG_ULEN));
#else
	/**
	  * @date Wed Jun 22 20:56:59 MDT 2005
	  *
	  * This no longer works on FC4 2.6.11 kernel: validity checks are
	  * performed on the length before we get here.  We might as well patch
	  * this out for all kernels and use the ioctl method instead.  Emulating
	  * an system call in this fashion was foolish in the first place: the
	  * normal method for system call emulation under UNIX is with ioctl. -bb
	  */
	struct __lis_putpmsg_args {
		int fd;
		struct strbuf *ctlptr;
		struct strbuf *datptr;
		int band;
		int flags;
	};
	struct __lis_putpmsg_args args = {
		fd:fd,
		ctlptr:ctlptr,
		datptr:datptr,
		band:band,
		flags:flags,
	};

	return (write(fd, &args, LIS_GETMSG_PUTMSG_ULEN));
#endif
}

/** @brief put a message to a stream band.
  * @param fd a file descriptor representing the stream.
  * @param ctlptr a pointer to a strbuf structure describing the control part of the message.
  * @param datptr a pointer to a strbuf structure describing the data part of the message.
  * @param band the band to which to put the message.
  * @param flags the priority of the message.
  */
int
__lis_putpmsg(int fd, struct strbuf *ctlptr, struct strbuf *datptr, int band, int flags)
{
#ifdef BLD32OVER64
	putpmsg_args6_t args;
	strbuf6_t ctl6;
	strbuf6_t *ptrc6 = 0;
	strbuf6_t dat6;
	strbuf6_t *ptrd6 = 0;

	if (ctlptr) {
		ctl6.maxlen = ctlptr->maxlen;
		ctl6.len = ctlptr->len;
		ctl6.buf = (unsigned long long) (unsigned long) ctlptr->buf;
		ptrc6 = &ctl6;
	}

	if (dataptr) {
		dat6.maxlen = dataptr->maxlen;
		dat6.len = dataptr->len;
		dat6.buf = (unsigned long long) (unsigned long) dataptr->buf;
		ptrd6 = &dat6;
	}

	memset((void *) &args, 0, sizeof(putpmsg_args6_t));
	args.fd = fd;
	args.ctl = (unsigned long long) (unsigned long) ptrc6;
	args.dat = (unsigned long long) (unsigned long) ptrd6;
	args.band = band;
	args.flags = flags;

	return (write(fd, &args, LIS_GETMSG_PUTMSG_ULEN));
#else
	struct strpmsg args = { {-1, -1, NULL}, {-1, -1, NULL}, band, flags };
	int rc;

	if (ctlptr)
		args.ctlbuf = *ctlptr;
	if (datptr)
		args.databuf = *datptr;
	if ((rc = ioctl(fd, I_PUTPMSG, &args)) != -1) {
		if (ctlptr)
			*ctlptr = args.ctlbuf;
		if (datptr)
			*datptr = args.databuf;
	}
	return (rc);
#endif
}

/** @brief put a message to a stream band.
  * @param fd a file descriptor representing the stream.
  * @param ctlptr a pointer to a strbuf structure describing the control part of the message.
  * @param datptr a pointer to a strbuf structure describing the data part of the message.
  * @param band the band to which to put the message.
  * @param flags the priority of the message.
  * @version LIS_1.0 putpmsg()
  *
  * putpmsg() must contain a thread cancellation point (SUS/XOPEN/POSIX).  In
  * the Linux Threads approach, this function will return EINTR if interrupted
  * by a signal.  When the function returns EINTR, the Linux Threads user should
  * check for cancellation with pthread_testcancel().  Because this function
  * consists of a single system call, asynchronous thread cancellation
  * protection is not required.
  */
int
__lis_putpmsg_r(int fd, struct strbuf *ctlptr, struct strbuf *datptr, int band, int flags)
{
	int ret;

	pthread_testcancel();
	ret = __lis_putpmsg(fd, ctlptr, datptr, band, flags);
	if (ret == -1)
		pthread_testcancel();
	return (ret);
}

/** @version LIS_0.0 putpmsg()
  */
int
__old_lis_putpmsg_r(int fd, struct strbuf *ctlptr, struct strbuf *datptr, int band, int flags)
{
	int ret;

	pthread_testcancel();
	ret = __old_lis_putpmsg(fd, ctlptr, datptr, band, flags);
	if (ret == -1)
		pthread_testcancel();
	return (ret);
}

/** @fn int putpmsg(int fd, struct strbuf *ctlptr, struct strbuf *datptr, int band, int flags)
  * @brief put a message to a stream band.
  * @param fd a file descriptor representing the stream.
  * @param ctlptr a pointer to a strbuf structure describing the control part of the message.
  * @param datptr a pointer to a strbuf structure describing the data part of the message.
  * @param band the band to which to put the message.
  * @param flags the priority of the message.
  * @version LIS_1.0 __lis_putpmsg_r()
  * @version LIS_0.0 __old_lis_putpmsg_r()
  *
  * This symbol is aliased to __lis_putpmsg_r().
  */
__asm__(".symver __lis_putpmsg_r,putpmsg@@LIS_1.0");
__asm__(".symver __old_lis_putpmsg_r,putpmsg@LIS_0.0");

/** @} */

// vim: com=srO\:/**,mb\:*,ex\:*/,srO\:/*,mb\:*,ex\:*/,b\:TRANS
